src/BitStringAddresses/sortedparticlelist.jl

```diff
@@ -0,0 +1,243 @@
+    end
+end
+
+function Base.reverse(ss::SortedParticleList{N,M,T}) where {N,M,T}
+    new_storage = map(reverse(ss.storage)) do i
+        T(M) - i + one(T)
+    end
+    return SortedParticleList{N,M,T}(new_storage)
+end
+
+# Somehow this is faster than the default method.
+Base.hash(ss::SortedParticleList, u::UInt) = hash(ss.storage, u)
+
+# In this case, getting the ONR is the same for bosons and fermions, and assumes the address
+# is not malformed.
+function onr(ss::SortedParticleList{<:Any,M}) where {M}
+    mvec = zeros(MVector{M,Int})
+    @inbounds for (occnum, mode, _) in ss
+        mvec[mode] = occnum
+    end
+    return SVector(mvec)
+end
+
+# Same as above.
+function find_mode(ss::SortedParticleList, n)
+    offset = 0
+    for (occnum, mode, _) in ss
+        if mode == n
+            return (occnum, mode, offset)
+        elseif mode > n
+            return (0, n, offset)
+        end
+        offset += occnum
+    end
+    return (0, n, offset)
+end
+
+"""
+    move_particles(ss::SortedParticleList, dsts, srcs)
+
+Move several particles at once. Moves `srcs[i]` to `dsts[i]`.  `dsts` and `srcs` should be
+tuples of [`BoseFSIndex`](@ref) or [`FermiFSIndex`](@ref). The legality of the moves is not
+checked - the result of an illegal move is undefined!
+"""
+function move_particles(ss::SortedParticleList{N,M,T}, dsts, srcs) where {N,M,T}
+    new_storage = ss.storage
+    for (dst, src) in zip(dsts, srcs)
+        src_pos = 1
+        @inbounds for i in 1:N
+            src_pos = max(src_pos, i * (new_storage[i] == src.mode % T))
+        end
+        @boundscheck 0 < dst.mode ≤ M || throw(BoundsError(ss, dst.mode))
+        new_storage = setindex(new_storage, dst.mode % T, src_pos)
+    end
+    return SortedParticleList{N,M,T}(sort(new_storage))
+end
+
+###
+### Bose interface
+###
+function from_bose_onr(::Type{S}, onr) where{S<:SortedParticleList}
+    from_onr(S, onr)
+end
+to_bose_onr(ss::SortedParticleList, _) = onr(ss)
+
+bose_num_occupied_modes(ss::SortedParticleList) = length(ss)
+
+bose_find_mode(ss::SortedParticleList, n) = find_mode(ss, n)
+
+@inline function bose_excitation(
+    ss::SortedParticleList{N,M,T}, creations, destructions
+) where {N,M,T}
+    creations_rev = reverse(creations)
+    value = bose_excitation_value(creations_rev, destructions)
+    if iszero(value)
+        return ss, 0.0
+    else
+        return move_particles(ss, creations_rev, destructions), √value
+    end
+end
+
+Base.length(bom::BoseOccupiedModes{<:Any,<:Any,<:SortedParticleList}) = length(bom.storage)
+function Base.iterate(bom::BoseOccupiedModes{<:Any,<:Any,<:SortedParticleList}, i=1)
+    it = iterate(bom.storage, i)
+    if isnothing(it)
+        return nothing
+    else
+        res, i = it
+        return BoseFSIndex(res...), i
+    end
+end
+
+###
+### Fermi interface
+###
+function from_fermi_onr(::Type{S}, onr) where {S<:SortedParticleList}
+    from_onr(S, onr)
+end
+
+function _fix_pos_create(c, index)
```

Comment:
Can we please have a bit of information around these functions `_fix_pos_create`, `_fix_pos_destroy`, `count_minus_signs`?

I understand that these are internal unexported functions that just used locally, so I think it is fine if they don't have docstrings. However, some code comments might be helpful to make the code more readable, as the code is not fully self-documenting.